### PR TITLE
Code Review - implementations/c

### PR DESCRIPTION
**Pull request by mrinalwadhwa**
THIS PR WILL NEVER BE MERGED.

### CODE REVIEW

File: implementations/c/documentation/examples/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.14)
```
Comment (mrinalwadhwa) [inline implementations/c/documentation/examples/CMakeLists.txt:1]: TODO: update examples for modern cmake

File: implementations/c/documentation/ockam/demo/handshake/include/handshake_test.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#include <stdlib.h>
```
Comment (mrinalwadhwa) [inline implementations/c/documentation/ockam/demo/handshake/include/handshake_test.h:1]: @robinbudd is the code in `implementations/c/documentation/ockam/demo/` still valid?
I suspect it isn't, we should either update to work with our latest code or remove.

File: implementations/c/ockam/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+add_subdirectory(queue)
+add_subdirectory(queue/pthread)
+
+add_subdirectory(mutex)
+add_subdirectory(mutex/pthread)
+
+add_subdirectory(vault)
+add_subdirectory(vault/default)
+# add_subdirectory(vault/atecc608a)
```
Comment (mrinalwadhwa) [inline implementations/c/ockam/CMakeLists.txt:22]: TODO: update vault/atecc608a to build with our latest cmake approach

File: implementations/c/ockam/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+# add_subdirectory(vault/atecc508a)
```
Comment (mrinalwadhwa) [inline implementations/c/ockam/CMakeLists.txt:23]: TODO: update vault/atecc608a to build with our latest cmake approach

File: implementations/c/ockam/channel/channel.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#ifndef OCKAM_CHANNEL_H
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel.h:1]: Create rules for how header guards look like (I think there already exist such rules, though I've seen different approaches throughout the project headers)

File: implementations/c/ockam/channel/channel.h [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+#define OCKAM_CHANNEL_H
+
+#include "ockam/error.h"
+#include "ockam/io.h"
+#include "ockam/memory.h"
+#include "ockam/vault.h"
+
+#define CHANNEL_ERROR_PARAMS          (OCKAM_ERROR_INTERFACE_CHANNEL | 0x0001u)
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel.h:9]: Consider using separate enum type for errors from specific module, this is both more readable (it's easier to find what the error is knowing it's type and code, rather that having some big integer and computing what's the error type bit & error code are) and eliminates need to have different non-repeatable error bit for each module (this involves excessive modules coupling)

File: implementations/c/ockam/channel/channel.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+#define CHANNEL_ERROR_NOT_IMPLEMENTED (OCKAM_ERROR_INTERFACE_CHANNEL | 0x0002u)
+#define CHANNEL_ERROR_KEY_AGREEMENT   (OCKAM_ERROR_INTERFACE_CHANNEL | 0x0003u)
+#define CHANNEL_ERROR_STATE           (OCKAM_ERROR_INTERFACE_CHANNEL | 0x0004u)
+
+typedef struct ockam_channel_t ockam_channel_t;
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel.h:14]: We need to finally make decision about headers structure and visibility

File: implementations/c/ockam/channel/channel.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+typedef struct ockam_channel_attributes_t {
+  ockam_reader_t* reader;
+  ockam_writer_t* writer;
+  ockam_memory_t* memory;
+  ockam_vault_t*  vault;
+} ockam_channel_attributes_t;
+
+ockam_error_t ockam_channel_init(ockam_channel_t* channel, ockam_channel_attributes_t* p_attrs);
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel.h:23]: Some of init functions use pointer to allocated structure, some takes pointer to pointer and allocates memory inside. Need to stick up to one way

File: implementations/c/ockam/channel/channel_impl.c [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+#include "ockam/memory.h"
+#include "ockam/key_agreement.h"
+#include "ockam/key_agreement/xx.h"
+#include "ockam/transport.h"
+#include "ockam/io/impl.h"
+#include "ockam/channel.h"
+#include "channel_impl.h"
+#include "ockam/codec.h"
+
+ockam_memory_t* gp_ockam_channel_memory = NULL;
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:15]: I would avoid having global variables. Each time channel instance is initiated this variable is overridden which is
1. Not obvious at all
2. Makes impossible using different memory implementations, which then makes all the complexity of having separate memory module with dispatch pointless, because when you have only 1 implementation at a time there are easier ways to achieve that.

File: implementations/c/ockam/channel/channel_impl.c [hunk 10]
```diff
@@ -0,0 +1,297 @@
+#include <stdlib.h>
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:1]: Check for unused headers

File: implementations/c/ockam/channel/channel_impl.c [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+ockam_error_t channel_write(void*, uint8_t*, size_t);
+
+uint8_t g_encoded_text[MAX_CHANNEL_PACKET_SIZE];
+uint8_t g_cipher_text[MAX_CHANNEL_PACKET_SIZE];
+
+uint8_t* channel_encode_header(ockam_channel_t* p_ch, uint8_t* p_encoded)
+{
+  p_encoded = encode_ockam_wire(p_encoded);
+  if (NULL == p_encoded) goto exit;
+  *p_encoded++ = 0; //!! onward route not implemented
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:27]: Add TODO or FIXME

File: implementations/c/ockam/channel/channel_impl.c [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+  *p_encoded++ = 0; //!! return route not implemented
+exit:
+  return p_encoded;
+}
+
+uint8_t* channel_deocde_header(ockam_channel_t* p_ch, uint8_t* p_encoded)
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:33]: typo

File: implementations/c/ockam/channel/channel_impl.c [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+{
+  p_encoded = decode_ockam_wire(p_encoded);
+  if (NULL == p_encoded) goto exit;
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:36]: Use either (p_encoded) and (!p_encoded) or (NULL != p_encoded) and (NULL == p_encoded)

File: implementations/c/ockam/channel/channel_impl.c [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+  if (!p_encoded) {
+    p_encoded = NULL;
+    goto exit;
+  }
+  if (0 != *p_encoded++) {
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:41]: Isn't it dead code?

File: implementations/c/ockam/channel/channel_impl.c [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+  *p_encoded++ = 0; //!! return route not implemented
+exit:
+  return p_encoded;
+}
+
+uint8_t* channel_deocde_header(ockam_channel_t* p_ch, uint8_t* p_encoded)
+{
+  p_encoded = decode_ockam_wire(p_encoded);
+  if (NULL == p_encoded) goto exit;
+  if (!p_encoded) {
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:37]: Always true

File: implementations/c/ockam/channel/channel_impl.c (hunk 12 above)
Comment (SanjoDeundiak) [reply implementations/c/ockam/channel/channel_impl.c:33]: p_ch is unused

File: implementations/c/ockam/channel/channel_impl.c [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+
+ockam_memory_t* gp_ockam_channel_memory = NULL;
+
+ockam_error_t channel_read(void*, uint8_t*, size_t, size_t*);
+ockam_error_t channel_write(void*, uint8_t*, size_t);
+
+uint8_t g_encoded_text[MAX_CHANNEL_PACKET_SIZE];
+uint8_t g_cipher_text[MAX_CHANNEL_PACKET_SIZE];
+
+uint8_t* channel_encode_header(ockam_channel_t* p_ch, uint8_t* p_encoded)
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:23]: p_ch is unused

Comment (SanjoDeundiak) [reply implementations/c/ockam/channel/channel_impl.c:23]: make static

File: implementations/c/ockam/channel/channel_impl.c [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+#include "ockam/channel.h"
+#include "channel_impl.h"
+#include "ockam/codec.h"
+
+ockam_memory_t* gp_ockam_channel_memory = NULL;
+
+ockam_error_t channel_read(void*, uint8_t*, size_t, size_t*);
+ockam_error_t channel_write(void*, uint8_t*, size_t);
+
+uint8_t g_encoded_text[MAX_CHANNEL_PACKET_SIZE];
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:20]: Not thread-safe

Comment (SanjoDeundiak) [reply implementations/c/ockam/channel/channel_impl.c:20]: Not thread-safe in terms that if we have more than one channel instances, using them from different threads is not safe. Also, using one channel instance from different threads is also unsafe, that may be OK, but at least deserves documenting (as well as all thread-safety/unsafety stuff in the library)

File: implementations/c/ockam/channel/channel_impl.c [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+  } //!!onward route
+  if (0 != *p_encoded++) {
+    p_encoded = NULL;
+    goto exit;
+  } //!!return route
+exit:
+  return p_encoded;
+}
+
+ockam_error_t channel_decrypt(ockam_channel_t* p_ch,
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:53]: Make all the file private functions static

File: implementations/c/ockam/channel/channel_impl.c [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+  return error;
+}
+
+ockam_error_t channel_process_message(uint8_t* p_encoded,
+                                      size_t   encoded_text_length,
+                                      uint8_t* p_clear_text,
+                                      size_t*  p_clear_text_length)
+{
+  ockam_error_t        error        = OCKAM_ERROR_NONE;
+  codec_message_type_t message_type = *p_encoded++;
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:81]: Move to separate function

File: implementations/c/ockam/channel/channel_impl.c [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+  if (error) log_error(error, __func__);
+  return error;
+}
+
+ockam_error_t ockam_channel_init(ockam_channel_t* p_ch, ockam_channel_attributes_t* p_attrs)
+{
+  ockam_error_t error = OCKAM_ERROR_NONE;
+
+  if ((NULL == p_ch) || (NULL == p_attrs) || (NULL == p_attrs->reader) || (NULL == p_attrs->writer) ||
+      (NULL == p_attrs->memory)) {
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:103]: p_attrs->vault?

File: implementations/c/ockam/channel/channel_impl.c [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+exit:
+  if (error) log_error(error, __func__);
+  return error;
+}
+
+ockam_error_t ockam_channel_init(ockam_channel_t* p_ch, ockam_channel_attributes_t* p_attrs)
+{
+  ockam_error_t error = OCKAM_ERROR_NONE;
+
+  if ((NULL == p_ch) || (NULL == p_attrs) || (NULL == p_attrs->reader) || (NULL == p_attrs->writer) ||
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:102]: I would change for asserts here

File: implementations/c/ockam/channel/channel_impl.c [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+
+  p_ch->transport_reader = p_attrs->reader;
+  p_ch->transport_writer = p_attrs->writer;
+
+  error = ockam_xx_key_initialize(
+    &p_ch->key, gp_ockam_channel_memory, p_ch->vault, p_ch->channel_reader, p_ch->channel_writer);
+
+  p_ch->state = CHANNEL_STATE_M1;
+
+exit:
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:129]: This could be achieved by (IMHO) much cleaner way by having several exit labels.

File: implementations/c/ockam/channel/channel_impl.c [hunk 23]
```diff
@@ -0,0 +1,297 @@
...
+  *p_writer = p_ch->channel_writer;
+
+exit:
+  if (error) log_error(error, __func__);
+  return error;
+}
+
+ockam_error_t channel_read(void* ctx, uint8_t* p_clear_text, size_t clear_text_size, size_t* p_clear_text_length)
+{
+  ockam_error_t    error               = 0;
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:171]: I would prefer to declare variables as much close to the point of usage as possible

File: implementations/c/ockam/channel/channel_impl.h [hunk 24]
```diff
@@ -0,0 +1,26 @@
+#ifndef CHANNEL_IMPL_H
+#define CHANNEL_IMPL_H
+
+#include "ockam/memory.h"
+#include "ockam/key_agreement.h"
+
+#define MAX_CHANNEL_PACKET_SIZE 0x7fffu
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.h:7]: Should not be here. What's the motivation about this const been this huge?

Comment (SanjoDeundiak) [reply implementations/c/ockam/channel/channel_impl.h:7]: What's the motivation about this const been this huge?

File: implementations/c/ockam/channel/channel_impl.c [hunk 25]
```diff
@@ -0,0 +1,297 @@
...
+                                      size_t*  p_clear_text_length)
+{
+  ockam_error_t        error        = OCKAM_ERROR_NONE;
+  codec_message_type_t message_type = *p_encoded++;
+  switch (message_type) {
+  case PING:
+    break;
+  case PAYLOAD:
+    *p_clear_text_length = encoded_text_length - sizeof(uint8_t);
+    ockam_memory_copy(gp_ockam_channel_memory, p_clear_text, p_encoded, *p_clear_text_length);
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:87]: Is there a way to avoid copy?

File: implementations/c/ockam/channel/channel_impl.c [hunk 26]
```diff
@@ -0,0 +1,297 @@
...
+
+  error = ockam_memory_alloc_zeroed(gp_ockam_channel_memory, (void**) &p_ch->channel_writer, sizeof(ockam_writer_t));
+  if (error) goto exit;
+  p_ch->channel_writer->write = channel_write;
+  p_ch->channel_writer->ctx   = p_ch;
+
+  p_ch->transport_reader = p_attrs->reader;
+  p_ch->transport_writer = p_attrs->writer;
+
+  error = ockam_xx_key_initialize(
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:124]: Should call interface function, not concrete implementation

File: implementations/c/ockam/channel/channel_impl.c [hunk 27]
```diff
@@ -0,0 +1,297 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <ctype.h>
+#include "ockam/syslog.h"
+#include "ockam/memory.h"
+#include "ockam/key_agreement.h"
+#include "ockam/key_agreement/xx.h"
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:8]: Should not be here

File: implementations/c/ockam/channel/channel_impl.c [hunk 28]
```diff
@@ -0,0 +1,297 @@
...
+#include "ockam/transport.h"
+#include "ockam/io/impl.h"
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:10]: Should not be here

File: implementations/c/ockam/channel/channel_impl.c [hunk 29]
```diff
@@ -0,0 +1,297 @@
...
+    goto exit;
+  }
+
+  if (CHANNEL_STATE_SECURE == p_ch->state) {
+    error = channel_process_message(p_encoded, encoded_text_length, p_clear_text, p_clear_text_length);
+    if (error) goto exit;
+  } else {
+    codec_message_type_t message_type = *p_encoded++;
+    *p_clear_text_length              = encoded_text_length - (p_encoded - g_encoded_text);
+    ockam_memory_copy(gp_ockam_channel_memory, p_clear_text, p_encoded, *p_clear_text_length);
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:196]: What is this copy for?

File: implementations/c/ockam/channel/channel_impl.c [hunk 30]
```diff
@@ -0,0 +1,297 @@
...
+    switch (p_ch->state) {
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:197]: It looks like this code relies on message order guarantee, but AFAIK underlying transport doesn't fulfil that requirement.

File: implementations/c/ockam/channel/channel_impl.c [hunk 31]
```diff
@@ -0,0 +1,297 @@
...
+  ockam_channel_t* p_ch                = (ockam_channel_t*) ctx;
+
+  p_encoded = channel_encode_header(p_ch, p_encoded);
+  if (!p_encoded) {
+    error = CHANNEL_ERROR_NOT_IMPLEMENTED;
+    goto exit;
+  }
+
+  if (CHANNEL_STATE_SECURE == p_ch->state) {
+    *p_encoded++        = PAYLOAD;
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:245]: Move to separate function. Probably doesn't work between be/le

File: implementations/c/ockam/channel/channel_impl.c [hunk 32]
```diff
@@ -0,0 +1,297 @@
...
+      break;
+    case CHANNEL_STATE_SECURE:
+      break;
+    default:
+      error = CHANNEL_ERROR_NOT_IMPLEMENTED;
+      goto exit;
+    }
+    encoded_text_length = p_encoded - g_encoded_text + clear_text_length;
+    cipher_text_length  = encoded_text_length;
+    ockam_memory_copy(gp_ockam_channel_memory, p_encoded, p_clear_text, clear_text_length);
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:273]: What is this copy for?

File: implementations/c/ockam/channel/channel_impl.c [hunk 33]
```diff
@@ -0,0 +1,297 @@
...
+    ockam_memory_copy(gp_ockam_channel_memory, g_cipher_text, g_encoded_text, encoded_text_length);
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:274]: What is this copy for?

File: implementations/c/ockam/channel/channel_impl.c [hunk 34]
```diff
@@ -0,0 +1,297 @@
...
+      *p_encoded++ = KEY_AGREEMENT_T1_M3;
+      p_ch->state  = CHANNEL_STATE_SECURE;
+      break;
+    case CHANNEL_STATE_SECURE:
+      break;
+    default:
+      error = CHANNEL_ERROR_NOT_IMPLEMENTED;
+      goto exit;
+    }
+    encoded_text_length = p_encoded - g_encoded_text + clear_text_length;
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.c:271]: Such arithmetic is extremely hard to follow for me

File: implementations/c/ockam/channel/channel.h [hunk 35]
```diff
@@ -0,0 +1,28 @@
...
+
+#define CHANNEL_ERROR_PARAMS          (OCKAM_ERROR_INTERFACE_CHANNEL | 0x0001u)
+#define CHANNEL_ERROR_NOT_IMPLEMENTED (OCKAM_ERROR_INTERFACE_CHANNEL | 0x0002u)
+#define CHANNEL_ERROR_KEY_AGREEMENT   (OCKAM_ERROR_INTERFACE_CHANNEL | 0x0003u)
+#define CHANNEL_ERROR_STATE           (OCKAM_ERROR_INTERFACE_CHANNEL | 0x0004u)
+
+typedef struct ockam_channel_t ockam_channel_t;
+
+typedef struct ockam_channel_attributes_t {
+  ockam_reader_t* reader;
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel.h:17]: Add key agreement

File: implementations/c/ockam/channel/tests/channel_test.c [hunk 36]
```diff
@@ -0,0 +1,140 @@
+#include <stdio.h>
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/tests/channel_test.c:1]: Need to be refactored like transport tests

Comment (SanjoDeundiak) [reply implementations/c/ockam/channel/tests/channel_test.c:1]: Add more tests, e.g. out-of-order transport case, some negative tests

Comment (SanjoDeundiak) [reply implementations/c/ockam/channel/tests/channel_test.c:1]: Consider thread-safety tests

Comment (SanjoDeundiak) [reply implementations/c/ockam/channel/tests/channel_test.c:1]: Consider removing forks

File: implementations/c/ockam/channel/channel.h (hunk 6 above)
Comment (mrinalwadhwa) [reply implementations/c/ockam/channel/channel.h:9]: Agree 💯  `ockam/error.h` is a source of a lot of coupling.
I'm not sure if we even need the `ockam_error_t` as a type.

Could you prototype with a quick example in one module what this may look like.

File: implementations/c/ockam/channel/channel_impl.h [hunk 37]
```diff
@@ -0,0 +1,26 @@
...
+
+typedef enum {
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.h:9]: This state corresponds to xx key agreement, channel should not rely on that.

Comment (SanjoDeundiak) [reply implementations/c/ockam/channel/channel_impl.h:9]: I guess key agreement interface change is required to fix that

File: implementations/c/ockam/channel/channel_impl.h [hunk 38]
```diff
@@ -0,0 +1,26 @@
...
+  CHANNEL_STATE_M1     = 1,
+  CHANNEL_STATE_M2     = 2,
+  CHANNEL_STATE_M3     = 3,
+  CHANNEL_STATE_SECURE = 4
+} channel_state_t;
+
+struct ockam_channel_t {
+  channel_state_t state;
+  ockam_reader_t* transport_reader;
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.h:18]: This header doesn't include anything that could have ockam_reader_t definition or forward declaration. Same for ockam_writer_t and ockam_vault_t.

File: implementations/c/ockam/channel/channel_impl.h [hunk 39]
```diff
@@ -0,0 +1,26 @@
...
+  ockam_writer_t* transport_writer;
+  ockam_reader_t* channel_reader;
+  ockam_writer_t* channel_writer;
+  ockam_vault_t*  vault;
+  ockam_key_t     key;
```
Comment (SanjoDeundiak) [inline implementations/c/ockam/channel/channel_impl.h:23]: Should become a pointer after moving to interface usage instead of concrete implementation

**Comment by mrinalwadhwa** [comment]
@SanjoDeundiak let's capture whatever we learnt from this PR as issues, link to any discussions that started here in the issue.
After that let's close this PR and instead do module by module reviews over the next few weeks.

**Comment by mrinalwadhwa** [comment]
@SanjoDeundiak I'm closing this, let's capture any notes as issues.
